File: 55.java
// Leetcode 55 Jump Game
// Reference: https://leetcode.com/problems/jump-game/solutions/3758301/java-easy-solution-with-explanation/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

class Solution {
    public boolean canJump(int[] nums) {

        int reachable = 0;
        for(int i = 0; i < nums.length; i++){
            if(i > reachable) return false;
            reachable = Math.max(reachable, i+nums[i]);
        }
        return true;
        
    }
}

File: 55.js
// Leetcode 55 Jump Game
// Reference: https://leetcode.com/problems/jump-game/solutions/273641/javascript-simple-o-n-greedy-solution/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
    let idx = 0;
    let max = 0;
    let target = nums.length - 1;

    while (idx < nums.length){
        max = Math.max(max, idx + nums[idx]);

        if (max >= target){
            return true;
        }

        // the early exit condition
        if (max <= idx && nums[idx] === 0){
            return false;
        }
        idx ++;
    }
    return false;
};